### PR TITLE
Fix AllocationPoll::newRun fails when requests exceed largestSizeClass

### PR DESCRIPTION
**Pull request by marin-ma**
When directly invoking AllocationPoll::newRun like code piece below:
https://github.com/facebookincubator/velox/blob/35ab8cc2fcb3da75b602cd6f3187314ded6507f2/velox/common/memory/HashStringAllocator.cpp#L154

If the value of param `needed` exceed largestSizeClass, the program fails with exception thrown 
```
E0423 11:29:24.637616 43478 Exceptions.h:68] Line: /home/rong/github/oap-project/velox/velox/common/memory/MemoryAllocator.cpp:51, Function:allocationSize, Expression: minSizeClass <= sizeClassSizes_.back() (257 vs. 256) Requesting minimum size 257 larger than largest size class 256, Source: RUNTIME, ErrorCode: INVALID_STATE
```

This is caused by passing the `numPages` as the `minSizeClass` and it fails the check here
https://github.com/facebookincubator/velox/blob/35ab8cc2fcb3da75b602cd6f3187314ded6507f2/velox/common/memory/MemoryAllocator.cpp#L51-L56

### CODE REVIEW

**Comment by marin-ma** [comment]
cc @zhejiangxiaomai

File: velox/common/memory/AllocationPool.cpp [hunk 1]
```diff
@@ -81,7 +81,7 @@ void AllocationPool::newRunImpl(memory::MachinePageCount numPages) {
           std::make_unique<memory::Allocation>(std::move(allocation_)));
     }
     pool_->allocateNonContiguous(
-        std::max<int32_t>(kMinPages, numPages), allocation_, numPages);
+        std::max<int32_t>(kMinPages, numPages), allocation_);
```
Comment (xiaoxmeng) [inline velox/common/memory/AllocationPool.cpp:84]:
```
pool_->allocateNonContiguous(
        std::max<int32_t>(kMinPages, numPages), allocation_, std::min(numPages, pool_->largestSizeClass()));
```

File: velox/common/memory/tests/MemoryAllocatorTest.cpp [hunk 2]
```diff
@@ -1194,6 +1194,14 @@ TEST_P(MemoryAllocatorTest, allocation) {
   allocation->clear();
 }
 
+TEST_P(MemoryAllocatorTest, exceedLargestSizeClass) {
+  const size_t kExceedLargestSizeClass = instance_->largestSizeClass() + 1;
```
Comment (xiaoxmeng) [inline velox/common/memory/tests/MemoryAllocatorTest.cpp:1198]: Can you test with different allocation size based on the class size and check the number of page runs in the resulting allocation? 
```
AllocationPool::allocationAt(0)?
1, => expect one page run?
first size class + 1 => expect two page runs?
second size class + 1 => expect two page runs?
...
largestSizeClass() + 1 => expect two page runs?
```
You can structure the test with a while loop and each iteration test different size + expected page run? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@marin-ma thanks for the fix!

File: velox/common/memory/tests/MemoryAllocatorTest.cpp (hunk 2 above)
Comment (marin-ma) [reply velox/common/memory/tests/MemoryAllocatorTest.cpp:1198]: Thanks for your comments! I tried to add these checks but I found in most cases `allocation->numRuns() == 1`. Is it as expected, or did I misunderstand the above logic? Please review the updated code.

File: velox/common/memory/tests/MemoryAllocatorTest.cpp [hunk 3]
```diff
@@ -1194,6 +1194,30 @@ TEST_P(MemoryAllocatorTest, allocation) {
...
+  AllocationPool pool(pool_.get());
+  const auto* allocation = pool.allocationAt(0);
+  pool.newRun(1 * AllocationTraits::kPageSize);
```
Comment (xiaoxmeng) [inline velox/common/memory/tests/MemoryAllocatorTest.cpp:1201]:
```
nit: AllocationTraits::pageBytes(numPages)
```

File: velox/common/memory/tests/MemoryAllocatorTest.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/MemoryAllocatorTest.cpp:1198]: I think the number of page runs is most likely 1. I previous comment on the expected page run is incorrect. But you need to update allocation object as newRun() will insert new run: AllocationPool::allocationAt(currentRunIndex()). Then it suppose all the newRun should have only one pageRun.

**Comment by marin-ma** [comment]
@xiaoxmeng Can we merge this PR?

**Review by xiaoxmeng — APPROVED** [review]
@marin-ma thanks for the update!

**Comment by xiaoxmeng** [comment]
> @marin-ma thanks for the fix!

@marin-ma I chat with @oerling about this. This is intentionally designed so. Can you provide the call stack that run into this issue? Thanks!

**Comment by marin-ma** [comment]
@xiaoxmeng Here's the stacktrace. It's triggered when the allocated size is equal to largeSizeClass (256), and more space is required because of allocating HashStringAllocator::Header. So the final request size is 257 pages. `facebook::velox::HashStringAllocator::allocate(int, bool)` 
```
 0# facebook::velox::memory::MemoryAllocator::allocationSize(unsigned long, unsigned long) const in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
 1# 0x00007FEF029A7E9C in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
 2# facebook::velox::memory::MemoryPoolImpl::allocateNonContiguous(unsigned long, facebook::velox::memory::Allocation&, unsigned long) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
 3# facebook::velox::AllocationPool::newRunImpl(unsigned long) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
 4# facebook::velox::HashStringAllocator::newSlab(int) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
 5# facebook::velox::HashStringAllocator::allocate(int, bool) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
 6# std::vector<unsigned long, facebook::velox::StlAllocator<unsigned long> >::_M_default_append(unsigned long) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
 7# facebook::velox::BloomFilter<facebook::velox::StlAllocator<unsigned long> >::merge(char const*) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
 8# facebook::velox::functions::sparksql::makeMightContain(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<facebook::velox::exec::VectorFunctionArg, std::allocator<facebook::velox::exec::VectorFunctionArg> > const&) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
 9# std::_Function_handler<std::shared_ptr<facebook::velox::exec::VectorFunction> (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<facebook::velox::exec::VectorFunctionArg, std::allocator<facebook::velox::exec::VectorFunctionArg> > const&), std::shared_ptr<facebook::velox::exec::VectorFunction> (*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<facebook::velox::exec::VectorFunctionArg, std::allocator<facebook::velox::exec::VectorFunctionArg> > const&)>::_M_invoke(std::_Any_data const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<facebook::velox::exec::VectorFunctionArg, std::allocator<facebook::velox::exec::VectorFunctionArg> > const&) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
10# facebook::velox::exec::getVectorFunction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::shared_ptr<facebook::velox::Type const>, std::allocator<std::shared_ptr<facebook::velox::Type const> > > const&, std::vector<std::shared_ptr<facebook::velox::BaseVector>, std::allocator<std::shared_ptr<facebook::velox::BaseVector> > > const&) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
11# 0x00007FEF027AF7BD in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
12# 0x00007FEF027ADAB7 in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
13# facebook::velox::exec::compileExpressions(std::vector<std::shared_ptr<facebook::velox::core::ITypedExpr const>, std::allocator<std::shared_ptr<facebook::velox::core::ITypedExpr const> > > const&, facebook::velox::core::ExecCtx*, facebook::velox::exec::ExprSet*, bool) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
14# facebook::velox::exec::ExprSet::ExprSet(std::vector<std::shared_ptr<facebook::velox::core::ITypedExpr const>, std::allocator<std::shared_ptr<facebook::velox::core::ITypedExpr const> > > const&, facebook::velox::core::ExecCtx*, bool) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
15# 0x00007FEF00A74440 in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
16# facebook::velox::connector::hive::HiveDataSource::HiveDataSource(std::shared_ptr<facebook::velox::RowType const> const&, std::shared_ptr<facebook::velox::connector::ConnectorTableHandle> const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::shared_ptr<facebook::velox::connector::ColumnHandle>, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::shared_ptr<facebook::velox::connector::ColumnHandle> > > > const&, facebook::velox::CachedFactory<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, facebook::velox::FileHandle, facebook::velox::FileHandleGenerator, facebook::velox::FileHandleSizer, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >*, facebook::velox::memory::MemoryPool*, facebook::velox::connector::ExpressionEvaluator*, facebook::velox::memory::MemoryAllocator*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, folly::Executor*) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
17# facebook::velox::connector::hive::HiveConnector::createDataSource(std::shared_ptr<facebook::velox::RowType const> const&, std::shared_ptr<facebook::velox::connector::ConnectorTableHandle> const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::shared_ptr<facebook::velox::connector::ColumnHandle>, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::shared_ptr<facebook::velox::connector::ColumnHandle> > > > const&, facebook::velox::connector::ConnectorQueryCtx*) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
18# facebook::velox::exec::TableScan::getOutput() in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
19# facebook::velox::exec::Driver::runInternal(std::shared_ptr<facebook::velox::exec::Driver>&, std::shared_ptr<facebook::velox::exec::BlockingState>&, std::shared_ptr<facebook::velox::RowVector>&) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
20# facebook::velox::exec::Driver::next(std::shared_ptr<facebook::velox::exec::BlockingState>&) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
21# facebook::velox::exec::Task::next(folly::SemiFuture<folly::Unit>*) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
22# gluten::WholeStageResultIterator::Next() in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
23# arrow::Result<std::shared_ptr<gluten::ColumnarBatch> > arrow::Iterator<std::shared_ptr<gluten::ColumnarBatch> >::Next<gluten::ResultIterator::Wrapper<gluten::WholeStageResultIteratorFirstStage> >(void*) in /home/sparkuser/github/oap-project/gluten/cpp/build/releases/libvelox.so
24# gluten::ResultIterator::GetNext() in ./generic_benchmark
```

**Comment by marin-ma** [comment]
@xiaoxmeng Do you have further comments? Can you provide the internal linter error for me to fix that?

**Comment by marin-ma** [comment]
@xiaoxmeng Sorry for not catching up this PR for long time. Do you still have further comments? We still need this fix.

**Comment by FelixYBW** [comment]
@xiaoxmeng @mbasmanova Any issue to follow up on the PR? It's a bug fix of a Gluten case

**Comment by mbasmanova** [comment]
@FelixYBW Would you please rebase?

**Comment by marin-ma** [comment]
Close this one since there are a lot of changes on main branch and we no longer need this PR.

**Pull request closed without merging**